### PR TITLE
doc/user/unit-testing.md: explain fake client and unit test example

### PR DESCRIPTION
**Pull request by estroz**
**Description of the change:** doc on using `controller-runtime`'s fake client library to mock API calls in unit tests.

**Motivation for the change:** users should be writing unit tests for custom operator code. The fake client is useful for doing so.

### CODE REVIEW

**Comment by estroz** [comment]
The unit test here is less thorough than that in [`operator-sdk-samples`](https://github.com/operator-framework/operator-sdk-samples/blob/master/memcached-operator/pkg/controller/memcached/memcached_controller_test.go#L25) for brevity's sake. I'm adding a link in this doc to that test if users want to see the unabridged version.

File: doc/user/unit-testing.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  "k8s.io/apimachinery/pkg/types"
+  "k8s.io/client-go/kubernetes/scheme"
+  "sigs.k8s.io/controller-runtime/pkg/client/fake"
+  "sigs.k8s.io/controller-runtime/pkg/reconcile"
+  logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+func TestMemcachedControllerDeploymentCreate(t *testing.T) {
+  // Set the logger to development mode for verbose logs.
+  logf.SetLogger(logf.ZapLogger(true))
```
Comment (hasbro17) [inline doc/user/unit-testing.md]: Probably remove this line as it's not the focus of this doc or test.

File: doc/user/unit-testing.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  // Check the result of reconciliation to make sure it has the desired state.
+  if !res.Requeue {
+    t.Error("reconcile did not requeue request as expected")
+  }
+
+  // Check if deployment has been created and has the correct size.
+  dep := &appsv1.Deployment{}
+  err = cl.Get(context.TODO(), req.NamespacedName, dep)
+  if err != nil {
+    t.Fatalf("get deployment: (%v)", err)
```
Comment (hasbro17) [inline doc/user/unit-testing.md]:
```suggestion
    t.Fatalf("failed to get memcached deployment after reconcile: (%v)", err)
```

File: doc/user/unit-testing.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  // watched resource .
+  req := reconcile.Request{
+    NamespacedName: types.NamespacedName{
+      Name:      name,
+      Namespace: namespace,
+    },
+  }
+  res, err := r.Reconcile(req)
+  if err != nil {
+    t.Fatalf("reconcile: (%v)", err)
```
Comment (hasbro17) [inline doc/user/unit-testing.md]:
```suggestion
    t.Fatalf("failed to reconcile: (%v)", err)
```

**Pull request merged**